mapper/cumulocity/c8y_translator_lib/src/serializer.rs

```diff
@@ -126,3 +126,171 @@ impl GroupedMeasurementVisitor for C8yJsonSerializer {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use assert_json_diff::*;
+    use serde_json::json;
+
+    use super::*;
+    use chrono::{offset::FixedOffset, DateTime, Local};
+    fn test_timestamp() -> DateTime<FixedOffset> {
+        let local_time_now: DateTime<Local> = Local::now();
+        local_time_now.with_timezone(local_time_now.offset())
+    }
+    #[test]
+    fn serialize_single_value_message() {
+        let mut serializer = C8yJsonSerializer::new().unwrap();
+        let timestamp = test_timestamp();
+        serializer.timestamp(timestamp).unwrap();
+        serializer.measurement("temperature", 25.5).unwrap();
+        let output = serializer.bytes().unwrap();
+
+        let expected_output = json!({
+            "type": "ThinEdgeMeasurement",
+            "time": timestamp.to_rfc3339(),
+            "temperature":{
+                "temperature":{
+                    "value": 25.5
+                }
+            }
+        });
+
+        assert_json_eq!(
+            serde_json::from_slice::<serde_json::Value>(&output).unwrap(),
+            expected_output
+        );
+    }
+    #[test]
+    fn serialize_multi_value_message() {
+        let mut serializer = C8yJsonSerializer::new().unwrap();
+        let timestamp = test_timestamp();
+        serializer.timestamp(timestamp).unwrap();
+        serializer.measurement("temperature", 25.5).unwrap();
+        serializer.start_group("location").unwrap();
+        serializer.measurement("alti", 2100.4).unwrap();
+        serializer.measurement("longi", 2200.4).unwrap();
+        serializer.measurement("lati", 2300.4).unwrap();
+        serializer.end_group().unwrap();
+        serializer.measurement("pressure", 255.2).unwrap();
+
+        let output = serializer.bytes().unwrap();
+
+        let expected_output = json!({
+            "type": "ThinEdgeMeasurement",
+            "time": timestamp.to_rfc3339(),
+            "temperature":{
+                "temperature":{
+                    "value": 25.5
+                }
+            },
+             "location": {
+                 "alti": {
+                     "value": 2100.4
+                 },
+                 "longi":{
+                     "value": 2200.4
+                 },
+                 "lati":{
+                     "value": 2300.4
+                 },
+             },
+             "pressure":{
+                 "pressure":{
+                     "value":255.2
+                 }
+             }
+
+        });
+
+        assert_json_eq!(
+            serde_json::from_slice::<serde_json::Value>(&output).unwrap(),
+            expected_output
+        );
+    }
+
+    #[test]
+    fn serialize_empty_message() {
+        let serializer = C8yJsonSerializer::new().unwrap();
+        let expected_output: Vec<u8> = format!(r#"{{"type": "ThinEdgeMeasurement"}}"#).into_bytes();
```

Comment:
why use `format!`?

Reply:
Use a raw byte literal and just `.into()`.

Reply:
removed format!, and using json! macro.